01-ruby-fundamentals/data-structures.md

```diff
@@ -1,211 +1,312 @@
 
 ### Arrays
 
-An array is a list of data and each item in the list can be accessed or *indexed* by a number.  The first element of any Array is index number 0.  You can always access the size of the Array with the length method.  
+Here are some facts about arrays.
+
+#### Creation
+
+- Arrays are created by putting square brackets `[]` around a list of things
+  - The list can be empty
+- Since an array is an object, we can store it in a variable
+- You can get the length of the array with `Array#length`
+- The array will keep track of the order elements were added
 
 ```ruby
-# declaring an Array
-my_list = [ "Bulbasaur", "Ivysaur", "Venusaur", "Charmander", "Charmeleon", "Charizard", "Squirtle", "Wartortle" ]
+students = ['Ada', 'Katherine']
 
-# Accessing the first element
-puts my_list[0]
+puts "#{students}"
+# => ['Ada', 'Katherine']
 
-# check the length of the array
-puts "My list has #{ my_list.length } elements"
+puts students.length
+# => 2
 ```
 
-**Question**: How can you find the last element of the Array?
+#### Indices and Elements
 
-You can loop through an array like this:
-```ruby
-deposits = [300, 250, 128, 64, 3016]
+- Each element in the array is assigned a number called its _index_
+  - Indices always start from 0
+  - This means the last element is at `arr.length - 1`
+  - Negative indices count backwards from the end, starting at `-1`
+- Any object can be an element in an array
+- You can add a new element to the array with the `Array#push` method or with `<<`
+- You can read elements using square brackets `[]` and the index
+- Changing the value of an element is done in the same way
+- Trying to read an element that doesn't exist will result in `nil`
 
-total = 0
-# in this loop n becomes each element of the array
-deposits.each do |n|
-	total += n
-end
+```ruby
+# Read the element at index 0
+puts students[0]
+# => Ada
+
+# Read the last element
+puts students[-1]
+# => Katherine
+
+# Add elements
+students.push('Grace')
+students << 'Edith'
+puts "#{students}"
+# => ['Ada', 'Katherine', 'Grace', 'Edith']
+
+puts students[2]
+# => 'Grace'
+
+# Change the element at index 2
+students[2] = 'Anita'
+puts students[2]
+# => 'Anita'
+
+puts "#{students}"
+# => ['Ada', 'Katherine', 'Anita', 'Edith']
+
+# Read an element not in the array
+puts students[10]
+# => nil
+
+puts students[4]
+# => nil
+# Why? Aren't there 4 elements?
 ```
-| Loop #  |   n   |   total   |
-|---------|-------|-----------|
-| 1       |  300  |    300    |
-| 2       |  250  |    550    |
-| 3       |  128  |    678    |
-| 4       |  64   |    742    |
-| 5       |  3016 |    3758   |
 
+#### Iteration
 
-This loop does the same thing with slightly different syntax.  
+- Iteration is the process of doing something to each element in the array, like printing it to the command line or adding 7 to it
+- To iterate manually, you can use a `times` loop on the length of the array
```

Comment:
Maybe describe what this does differently (using the index) rather than using "manually"